lib/checkfunctions.cpp

```diff
@@ -279,6 +280,70 @@ void CheckFunctions::checkIgnoredReturnValue()
     }
 }
 
+void CheckFunctions::checkIteratorTypeMismatch()
+{
+    const SymbolDatabase *symbolDatabase = mTokenizer->getSymbolDatabase();
+    for (const Scope *scope : symbolDatabase->functionScopes) {
+        for (const Token* tok = scope->bodyStart->next(); tok != scope->bodyEnd; tok = tok->next()) {
+            if (!Token::Match(tok, "%name% ( !!)"))
+                continue;
+
+            if (tok->str() != "for") {
```

Comment:
I don't understand why you explicitly only write such warning if the assignment is in the first expression of a for loop.

it seems to me that you can remove a bunch of code and then warn about this also:

    void foo(std::multimap<int, double>& mm) {
        std::map<int, double>::iterator it = mm.begin();
    }